### DISCUSSION TITLE
Intercepting msw calls in cypress

### DISCUSSION THREAD

**Discussion by d-ivashchuk** [mswjs/msw#805] [Q&A]
Hey @kettanaito! I've finally been able to use your amazing library and I must say it's just fabulous DX!

I can't get something working though. I have setup couple of handlers for couple of the routes:

```
 rest.get("/api/videos", (req, res, ctx) => {
    return res(ctx.json(videosMock));
  }),
  rest.get("/api/courses", (req, res, ctx) => {
    return res(ctx.json(coursesMock));
  }),
  rest.get("/api/purchases", (req, res, ctx) => {
    return res(ctx.json(purchasesMock));
  }),
```

And everything works perfectly, but I want to intercept these calls in cypress. Using something like this:

```
cy.visit("/my-courses");
    cy.intercept("GET", "/api/videos").as("getVideos");
    cy.wait("@getVideos");
```

No luck so far, it just doesn't react although I have the call logs from MSW in console within cypress browser. Any ideas?

**Comment by kettanaito** [comment]
Hey, @d-ivashchuk. Thank you for the feedback!

As far as I know `cy.intercept` stubs native XMLHttpRequest/fetch and stores the requests that match your predicate (`"GET", "/api/videos"). This stubbing part means that requests never really happen—they get "silenced" by Cypress in order to be mocked using their API. 

MSW doesn't encourage stubbing both conceptually and implementation-wise, so for a request to be captured by MSW it _must_ actually happen (you should see it in the "Network" tab of your dev tools just as any other request). Service Worker API can only intercept actual requests. 

I'd suggest dropping `cy.intercept` and let your application perform requests as it would in production, giving you more confidence in your tests. There's also a [React REST API example](https://github.com/mswjs/examples/blob/d6b6da65f4987b7049db0976fa6af50fafd514b1/examples/rest-react/e2e/Login.e2e.test.js) that features Cypress for E2E tests. The API mocking setup is done on the application's side (which is how you may use it for development already) and requires no additional configurations for Cypress.

**Reply by d-ivashchuk** [reply]
This answers my question perfectly, thanks! But what if I want to still listen to When request is complete? The whole intercept was mainly for that purpose honestly.

**Reply by kettanaito** [reply]
Most of the requests have their reflection in the UI. I recommend awaiting UI elements that are related to the request's completion, such as those that render something based on the response. 

In other cases, you can use [life-cycle events API](https://mswjs.io/docs/extensions/life-cycle-events) and await a specific request's completion (also this is shouldn't be a go-to solution):

```js
await new Promise((resolve) => {
  worker.on('request:end', (request) => {
    if (isRequestINeed(request) {
      resolve()
    }
  })
})
```

> You can get access to the `worker` by setting it globally on `window` for tests only.

**Reply by idolize** [reply]
I responded to this earlier issue where I was asking for something similar: https://github.com/mswjs/msw/issues/548#issuecomment-879316394

But the gist is that the Cypress docs [do encourage you to use `cy.request` as part of the test itself for things like seeding data or short-circuiting your UI layer](https://docs.cypress.io/guides/getting-started/testing-your-app#Bypassing-your-UI) and not just as a means of returning mock data to the UI. Having to get a reference to `window.fetch` seems like a bit of a hack, and I'd personally prefer to keep my tests clean and just move my mocks to a separate node process.

@kettanaito would you consider at least making it easier to run msw mocks in a separate node process (instead of via a service worker) if users do want to do things like this? The msw mocks themselves are very descriptive and would be a great source of truth to target (in my opinion) for all use cases, but right now I'm considering moving my team to something else due to this limitation.

**Reply by kettanaito** [reply]
That is a good recipe, but you can use other means to seed your state outside of UI. It really comes down to the testing setup you wish to have, and the testing level you target. 

> would you consider at least making it easier to run msw mocks in a separate node process

I'm 100% open to extensions. This is not something we will ship in the core library, but it makes perfect sense as an extension. Would you be interested in implementing it? 

I can imagine it being a function of handlers that establishes a local HTTP server:

```js
// local-server.js
import { createServer } from '@mswjs/http-server'
improt { handlers } from './handlers'

const server = createServer(handlers)
server.listen(9090, () => console.log('Ready at http://localhost:9090'))
```

As you said, handlers are a great source, and you can map them to conventional API routes in any framework of your choosing (i.e. ExpressJS) and have a functioning local server. 

You may want to look into the [request matching algorithm](https://github.com/mswjs/msw/blob/ff6c1c016faac8648c61eaca51dca46693a6374c/src/utils/getResponse.ts#L18) (and [here](https://github.com/mswjs/msw/blob/ff6c1c016faac8648c61eaca51dca46693a6374c/src/utils/handleRequest.ts#L59)), as it would be easier to establish a single permissive route on the server that does request matching + response resolution under the hood:

```js
import { getResponse } from 'msw'

app.use(async (req, res) => {
  const { response } = await getResponse(req, handlers)
  // ...similar logic to "handleRequest.ts" here
})
```

I'd be happy to support you in the process of writing this extension, provide codebase guidance and ensure that MSW core meets the demands of this extension.

**Reply by idolize** [reply]
Yes - I've already got this functioning (see [here](https://github.com/mswjs/msw/issues/548#issuecomment-762679369)) but had to copy/paste the `getResponse` matching algorithm - would you consider exporting that from the library? If so, I'd be happy to contribute something that could be used as both a connect/express middleware as well as a basic server (if you don't want to bother making your own and attaching the middleware).

What do you think @kettanaito?

**Reply by kettanaito** [reply]
I think this makes perfect sense as a middleware 👍🏻  Also don't mind exporting `getResponse` publicly, you can go ahead with this direction, it looks promising. If you wish, I can create a private repo under MSW org where we can collaborate on the solution before releasing it.

**Reply by idolize** [reply]
@kettanaito awesome! yes, feel free to create the repo and I'll push something to a branch you can take a look at and feel free to comment or modify as you see fit.

**Reply by kettanaito** [reply]
@idolize, I've created the [http-middleware](https://github.com/mswjs/http-middleware) repository with the minimal development setup. Looking forward to collaborating with you on this one!

**Reply by idolize** [reply]
@kettanaito I've created a PR here https://github.com/mswjs/http-middleware/pull/1 - let's move the discussion there!
